File: day_3/Python/rucksack_reorganization/main.py
def part_1(input_str):
    total = 0
    rucksacks = input_str.split("\n")

    for rucksack in rucksacks:
        size = len(rucksack) // 2

        char = ''

        left, right = rucksack[size:], rucksack[:size]
        for ch in left:
            if ch in right:
                char = ch

        total += ord(char) - 96 if char.islower() else ord(char) - 64 + 26

    print(total)


def main(input_str):
    part_1(input_str)


if __name__ == "__main__":
    main(
            )

File: day_3/Rust/rucksack_reorganization/src/main.rs
use std::collections::HashSet;

fn part_1(input_str: &str) -> i32 {
    let rucksacks: Vec<&str> = input_str.split("\n").collect::<Vec<&str>>();

    let mut total: i32 = 0;

    for rucksack in rucksacks {
        let size: usize = rucksack.len();

        let rucksack = rucksack
            .chars()
            .collect::<Vec<char>>();
         
        let left   = rucksack[..size/2].to_vec();
        let right  = rucksack[size/2..].to_vec();

        // uhh
        let left_set: HashSet<char> = left.iter().collect::<String>().chars().collect();
        let right_set: HashSet<char> = right.iter().collect::<String>().chars().collect();

        let ch = left_set
            .iter()
            .filter(
                |c| right_set
                    .contains(c)
            )
            .collect::<String>();


        let ch: char = ch.chars().nth(0).unwrap();
        total += match true {
            a if a == ch.is_lowercase() => ch as i32 - 96,
            b if b == ch.is_uppercase() => ch as i32 - 64 + 26,
            _ => panic!("bruh"),
        };
    }
    total
}

// fn part_2(input_str: String) -> i64 {
// }


fn main() {
    println!("{}", part_1(&input_str));
    // part_2(&input_str);
}
